src/main/scala/ai/tripl/arc/plugins/pipeline/ControlFlowExecute.scala

```diff
@@ -0,0 +1,162 @@
+        stage.stageDetail.put("params", params.asJava)
+
+        Right(stage)
+      case _ =>
+        val allErrors: Errors = List(name, key, description, parsedURI, inputSQL, validSQL, invalidKeys).collect{ case Left(errs) => errs }.flatten
+        val stageName = stringOrDefault(name, "unnamed stage")
+        val err = StageError(index, stageName, c.origin.lineNumber, allErrors)
+        Left(err :: Nil)
+    }
+  }
+}
+
+case class ControlFlowExecuteStage(
+    plugin: ControlFlowExecute,
+    name: String,
+    key: String,
+    description: Option[String],
+    inputURI: URI,
+    sql: String,
+    sqlParams: Map[String, String],
+    params: Map[String, String]
+  ) extends PipelineStage {
+
+  override def execute()(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
+    ControlFlowExecuteStage.execute(this)
+  }
+}
+
+case class ControlFlowPayload(
+  outcome: Boolean,
+  message: String
+)
+
+object ControlFlowExecuteStage {
+
+  def execute(stage: ControlFlowExecuteStage)(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
+
+    val signature = "ControlFlowExecute requires query to return 1 row with [outcome: boolean, message: string] signature."
+
+    val stmt = SQLUtils.injectParameters(stage.sql, stage.sqlParams, false)
+    stage.stageDetail.put("sql", stmt)
+
+    val df = try {
+      spark.sql(stmt)
+    } catch {
+      case e: Exception => throw new Exception(e) with DetailException {
+        override val detail = stage.stageDetail
+      }
+    }
+    val count = df.persist(arcContext.storageLevel).count
+
+    if (df.count != 1 || df.schema.length != 2) {
+      throw new Exception(s"""${signature} Query returned ${count} rows of type [${df.schema.map(f => f.dataType.simpleString).mkString(", ")}].""") with DetailException {
+        override val detail = stage.stageDetail
+      }
+    }
+
+
+    try {
+      val row = df.first
+      val resultIsNull = row.isNullAt(0)
+      val messageIsNull = row.isNullAt(1)
+
+      if (resultIsNull) {
+        throw new Exception(s"""${signature} Query returned [null, ${if (messageIsNull) "null" else "not null"}].""") with DetailException {
+          override val detail = stage.stageDetail
+        }
+      }
+
+      val message = row.getString(1)
+
+      // try to parse to json
+      try {
+        val objectMapper = new ObjectMapper()
+        var messageMap = new java.util.HashMap[String, Object]()
+        messageMap = objectMapper.readValue(message, classOf[java.util.HashMap[String, Object]])
+        stage.stageDetail.put("message", messageMap)
+      } catch {
+        case e: Exception =>
+          stage.stageDetail.put("message", message)
+      }
+
+      val result = row.getBoolean(0)
+
+      arcContext.userData.put(stage.key, ControlFlowPayload(result, message))
```

Comment:
If you've already checked to see if the message is JSON it would be good to retain that information in the ControlFlowPayload so that lifecycle plugins don't have to repeat the check. I suggest an ADT for string message vs JSON message (map in this case).

Reply:
moved into the ControlFlowPayload case class so logic is no longer repeated